stablehlo/reference/ProcessGrid.cpp

```diff
@@ -149,11 +149,35 @@ std::mutex &ProcessGrid::getRendezvousLock(ProcessGroup processGroup,
   return channelLocks_[channelKey];
 }
 
+std::mutex &ProcessGrid::getSendRecvOmniLock(ChannelId channelId) {
+  std::lock_guard<std::mutex> lock(rendezvousLock_);
+  return sendRecvChannelLocks_[channelId];
+}
+
 void ProcessGrid::outfeed(ArrayRef<Tensor> inputs) {
   std::lock_guard<std::mutex> lock(outfeedLock_);
   outfeed_.emplace(inputs);
 }
 
+SmallVector<Tensor> ProcessGrid::recv(ChannelId channelId,
+                                      ProcessId processId) {
+  {
+    std::lock_guard<std::mutex> lock(getSendRecvOmniLock(channelId));
+    if (sendRecvChannels_[channelId].size() != 0)
+      sendRecvChannels_[channelId].clear();
+    sendRecvReady_.insert(channelId);
+  }
+  sendRecvConditions_[channelId].notify_all();
+  {
+    std::unique_lock<std::mutex> lock(getSendRecvOmniLock(channelId));
+    if (!sendRecvConditions_[channelId].wait_for(
+            lock, std::chrono::seconds(3),
+            [&] { return sendRecvChannels_[channelId].size() != 0; }))
+      llvm::report_fatal_error("recv timed out");
+  }
+  return sendRecvChannels_[channelId].getSortedTensors();
```

Comment:
Should this access be lock-guarded? Same for `sendRecvConditions_[channelId].notify_all();` Will need to think more about these unguarded accesses of non-thread-safe datatypes.

Reply:
Rebased to new commit with a few changes.